Sasquatch/Sasquatch/MSAAnalyticsAuthenticationProvider.swift

```
@@ -0,0 +1,91 @@
+            request.httpMethod = "POST"
+            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
+            let bodyString = kMSARedirectParam.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)! + kMSAClientIdParam + kMSARefreshParam + refreshToken + kMSAScopeParam
+            let data: Data = bodyString.data(using: String.Encoding.utf8)!
+            
+            NSLog("Started refresh process")
+            session.uploadTask(with: request as URLRequest, from: data) { (data, response, error) in
+                defer {
+                    self.close()
+                }
+                do {
+                    guard let data = data else {
+                        
+                        // Call the completion handler in the error case to send anonymous logs.
+                        completionHandler(nil, nil)
+                        throw JSONError.NoData
+                    }
+                    guard let json = try JSONSerialization.jsonObject(with: data, options: []) as? NSDictionary else {
+                        
+                        // Call the completion handler in the error case to send anonymous logs.
+                        completionHandler(nil, nil)
+                        throw JSONError.ConversionFailed
+                    }
+                    if let error = json["error"] as? String, let errorDescription = json["error_description"] as? String {
+                        NSLog("Refresh token error: \"\(error)\": \(errorDescription)")
+                        
+                        // Call the completion handler in the error case to send anonymous logs.
+                        completionHandler(nil, nil)
+                        return
+                    }
+                    let token = json["access_token"]! as! String
+                    let expiresIn = json["expires_in"]! as! Int64
+                    let userId = json["user_id"]! as! String
```

let's avoid force unwrapping during json parsing

---

This is not my code 😒 
And I guess I can see the intention why it was written. We expect this value to be here in any case. If there are not, it's an error so better fail fast.